xds/internal/balancer/clusterimpl/picker.go

```diff
@@ -88,6 +88,11 @@ type picker struct {
 }
 
 func (b *clusterImplBalancer) newPicker(config *dropConfigs) *picker {
+	b.childStateMu.Lock()
```

Comment:
I would get rid of this method here, and would instead add a `newPickerLocked` in `clusterimpl.go` that is always called with the mutex held. In the process I would also get rid of the `dropConfigs` struct and would change `handleDropAndRequestCount()` to instead return a boolean.

```golang
// handleDropAndRequestCount compares drop and request counter in newConfig with
// the one currently used by picker. It returns a boolean indicating if a new
// picker needs to be generated.
func (b *clusterImplBalancer) handleDropAndRequestCount(newConfig *LBConfig) bool {
	var updatePicker bool
	if !equalDropCategories(b.dropCategories, newConfig.DropCategories) {
		b.dropCategories = newConfig.DropCategories
		b.drops = make([]*dropper, 0, len(newConfig.DropCategories))
		for _, c := range newConfig.DropCategories {
			b.drops = append(b.drops, newDropper(c))
		}
		updatePicker = true
	}

	if b.requestCounterCluster != newConfig.Cluster || b.requestCounterService != newConfig.EDSServiceName {
		b.requestCounterCluster = newConfig.Cluster
		b.requestCounterService = newConfig.EDSServiceName
		b.requestCounter = xdsclient.GetClusterRequestsCounter(newConfig.Cluster, newConfig.EDSServiceName)
		updatePicker = true
	}
	var newRequestCountMax uint32 = 1024
	if newConfig.MaxConcurrentRequests != nil {
		newRequestCountMax = *newConfig.MaxConcurrentRequests
	}
	if b.requestCountMax != newRequestCountMax {
		b.requestCountMax = newRequestCountMax
		updatePicker = true
	}

	return updatePicker
}

func (b *clusterImplBalancer) newPickerLocked() *picker {
	return &picker{
		drops:           b.drops,
		s:               b.childState,
		loadStore:       b.loadWrapper,
		counter:         b.requestCounter,
		countMax:        b.requestCountMax,
		telemetryLabels: b.telemetryLabels,
	}
}
```

Reply:
Updated, thAnks